File: tests/metrics/annualized_return_test.py
"""Correctness and lifecycle tests for AnnualizedReturn."""
from __future__ import annotations
import empyrical
import numpy as np
import pytest
from taflow.metrics import AnnualizedReturn

@pytest.mark.parametrize('periods_per_year', [1.0, 12.0, 52.0, 252.0, 365.0])
@pytest.mark.parametrize('returns', [np.array([0.02]), np.array([0.01, -0.02, 0.03, 0.0, -0.01]), np.zeros(32), np.linspace(-0.02, 0.03, 64), np.random.default_rng(9721).normal(0.0004, 0.012, 257), np.array([np.nan, 0.01, -0.02, np.nan, 0.015])])
def test_matches_empyrical_annual_return(returns: np.ndarray, periods_per_year: float) -> None:
    actual = AnnualizedReturn(periods_per_year=periods_per_year).from_returns(returns).compute()
    oracle_returns = returns[~np.isnan(returns)]
    expected = empyrical.annual_return(oracle_returns, annualization=periods_per_year)
    assert actual == pytest.approx(expected, rel=1e-12, abs=1e-14)

def test_input_methods_and_lifecycle_are_invariant() -> None:
    returns = np.array([0.1, -0.2, 0.05, 0.03], dtype=np.float64)
    equity = 100.0 * np.cumprod(np.r_[1.0, 1.0 + returns])
    pnl = np.diff(equity)
    log_returns = np.log1p(returns)
    expected = empyrical.annual_return(returns, annualization=12.0)
    states = [AnnualizedReturn(periods_per_year=12.0).from_returns([]), AnnualizedReturn(periods_per_year=12.0).from_log_returns(log_returns), AnnualizedReturn(periods_per_year=12.0).from_equity(equity), AnnualizedReturn(periods_per_year=12.0).from_pnl(pnl, initial_capital=100.0)]
    states[0].append(returns[0]).extend(returns[1:3]).append(returns[3])
    for state in states:
        assert state.compute() == pytest.approx(expected, rel=1e-12, abs=1e-14)
        assert state.value == state.compute()
        assert len(state) == returns.size
    streamed = states[0]
    assert streamed.reset() is streamed
    assert streamed.value is None
    assert streamed.compute() is None
    assert len(streamed) == 0
    assert streamed.extend(returns) is streamed
    assert streamed.compute() == pytest.approx(expected, rel=1e-12, abs=1e-14)

def test_nan_policy_and_edge_contract() -> None:
    omitted = AnnualizedReturn(periods_per_year=252.0).from_returns([np.nan, 0.25, -1.0])
    assert omitted.compute() == -1.0
    assert len(omitted) == 2
    with pytest.raises(ValueError, match='NaN'):
        AnnualizedReturn(nan_policy='raise').from_returns([0.01, np.nan])
    with pytest.raises(ValueError, match='periods_per_year'):
        AnnualizedReturn(periods_per_year=0.0).from_returns([])
    with pytest.raises(ValueError, match='simple returns'):
        AnnualizedReturn().from_returns([-1.01])
    with pytest.raises(ValueError, match='initial_capital'):
        AnnualizedReturn().from_pnl([1.0], initial_capital=0.0)

File: crates/taflow-metrics/src/metrics/annualized_return_test.rs
use super::annualized_return::AnnualizedReturn;
use crate::{MetricInputKind, NanPolicy};

fn assert_close(actual: f64, expected: f64) {
    assert!((actual - expected).abs() <= 1e-13, "{actual} != {expected}");
}

fn expected_annualized(returns: &[f64], periods_per_year: f64) -> f64 {
    let product = returns.iter().map(|value| 1.0 + value).product::<f64>();
    product.powf(periods_per_year / returns.len() as f64) - 1.0
}

#[test]
fn computes_geometric_annualized_return_and_preserves_lifecycle() {
    let returns = [0.10, -0.20, 0.05, 0.03];
    let expected = expected_annualized(&returns, 12.0);
    let mut state = AnnualizedReturn::new(12.0, NanPolicy::Omit).unwrap();
    state.from_returns(&[]).unwrap();

    assert_eq!(state.value(), None);
    assert_eq!(state.compute(), None);
    state.extend(&returns[..2]).unwrap();
    state.append(returns[2]).unwrap();
    assert_close(state.append(returns[3]).unwrap().unwrap(), expected);
    assert_close(state.compute().unwrap(), expected);
    assert_eq!(state.len(), 4);
    assert_eq!(state.periods_per_year(), 12.0);

    state.reset();
    assert!(state.is_empty());
    assert_eq!(state.value(), None);
    assert_close(state.extend(&returns).unwrap().unwrap(), expected);
}

#[test]
fn all_return_input_modes_are_equivalent() {
    let returns = [0.10, -0.20, 0.05];
    let expected = expected_annualized(&returns, 252.0);

    let mut equity = AnnualizedReturn::new(252.0, NanPolicy::Omit).unwrap();
    equity.from_equity(&[]).unwrap();
    assert_close(
        equity.extend(&[100.0, 110.0, 88.0, 92.4]).unwrap().unwrap(),
        expected,
    );
    assert_eq!(equity.len(), 3);

    let mut pnl = AnnualizedReturn::new(252.0, NanPolicy::Omit).unwrap();
    pnl.from_pnl(&[], 100.0).unwrap();
    assert_close(pnl.extend(&[10.0, -22.0, 4.4]).unwrap().unwrap(), expected);

    let log_returns = returns.map(f64::ln_1p);
    let mut logarithmic = AnnualizedReturn::new(252.0, NanPolicy::Omit).unwrap();
    logarithmic.from_log_returns(&[]).unwrap();
    assert_close(logarithmic.extend(&log_returns).unwrap().unwrap(), expected);
}

#[test]
fn omits_nan_and_represents_total_loss() {
    let mut state = AnnualizedReturn::new(252.0, NanPolicy::Omit).unwrap();
    state.from_returns(&[]).unwrap();
    state.extend(&[f64::NAN, 0.25, -1.0]).unwrap();
    assert_eq!(state.value(), Some(-1.0));
    assert_eq!(state.len(), 2);
}

#[test]
fn validates_configuration_and_semantic_domain() {
    for invalid in [0.0, -1.0, f64::NAN, f64::INFINITY] {
        assert!(AnnualizedReturn::new(invalid, NanPolicy::Omit).is_err());
    }
    let mut unbound = AnnualizedReturn::new(252.0, NanPolicy::Omit).unwrap();
    assert!(unbound.append(0.01).is_err());
}
